src/Middleware/RequestDecompression/src/DeflateDecompressionProvider.cs

```diff
@@ -13,6 +14,221 @@ internal sealed class DeflateDecompressionProvider : IDecompressionProvider
+        // use DeflateStream. This heuristic works because we're deciding only between raw deflate and zlib wrapped around
+        // deflate, in which case the first nibble will always be 8 for zlib and never be 8 for deflate.
+        // https://stackoverflow.com/a/37528114 provides an explanation for why.
+
+        public override int Read(Span<byte> buffer)
+        {
+            if (_decompressionStream is null)
+            {
+                int firstByte = _stream.PeekFirstByte();
+                _decompressionStream = CreateDecompressionStream(firstByte, _stream);
+            }
+
+            return _decompressionStream.Read(buffer);
+        }
+
+        public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+        {
+            if (_decompressionStream is null)
+            {
+                return CreateAndReadAsync(this, buffer, cancellationToken);
+
+                static async ValueTask<int> CreateAndReadAsync(ZLibOrDeflateStream thisRef, Memory<byte> buffer, CancellationToken cancellationToken)
+                {
+                    int firstByte = await thisRef._stream.PeekFirstByteAsync(cancellationToken).ConfigureAwait(false);
+                    thisRef._decompressionStream = CreateDecompressionStream(firstByte, thisRef._stream);
+                    return await thisRef._decompressionStream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return _decompressionStream.ReadAsync(buffer, cancellationToken);
+        }
+
+        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            ValidateCopyToArguments(destination, bufferSize);
+            return Core(destination, bufferSize, cancellationToken);
+            async Task Core(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            {
+                if (_decompressionStream is null)
+                {
+                    int firstByte = await _stream.PeekFirstByteAsync(cancellationToken).ConfigureAwait(false);
+                    _decompressionStream = CreateDecompressionStream(firstByte, _stream);
+                }
+
+                await _decompressionStream.CopyToAsync(destination, bufferSize, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        public override int Read(byte[] buffer, int offset, int count) => _stream.Read(buffer, offset, count);
+        public override long Seek(long offset, SeekOrigin origin) => _stream.Seek(offset, origin);
+        public override void SetLength(long value) => _stream.SetLength(value);
+        public override void Flush() => _stream.Flush();
+
+        private static Stream CreateDecompressionStream(int firstByte, Stream stream) =>
+            (firstByte & 0xF) == 8 ?
+                new ZLibStream(stream, CompressionMode.Decompress) :
```

Comment:
the ZLibStream and DeflateStream are getting passed a stream that's already advanced past the first byte here, right? are they OK with that?

Reply:
They're passed the peek stream that hasn't consumed the first byte. That's why the peek stream exists. The alternative would be to manual handle the header/footer.